packages/repository/src/__tests__/integration/repositories/relation.factory.integration.ts

```diff
@@ -203,6 +210,247 @@ describe('BelongsTo relation', () => {
   }
 });
 
+describe('HasManyThrough relation', () => {
+  let existingCustomerId: number;
+
+  let hasManyThroughRepo: HasManyThroughRepository<
+    Order,
+    typeof Order.prototype.id,
+    CustomerOrderLink
+  >;
+  let hasManyThroughFactory: HasManyThroughRepositoryFactory<
+    Order,
+    typeof Order.prototype.id,
+    CustomerOrderLink,
+    typeof Customer.prototype.id
+  >;
+
+  before(givenCrudRepositories);
+  before(givenPersistedCustomerInstance);
+  before(givenConstrainedRepositories);
+
+  beforeEach(async function resetDatabase() {
+    await customerRepo.deleteAll();
+    await customerOrderLinkRepo.deleteAll();
+    await orderRepo.deleteAll();
+  });
+
+  it('can create an target instance alone with the corresponding through model', async () => {
+    const order = await hasManyThroughRepo.create(
+      {
+        description: 'an order hasManyThrough',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    const persistedOrder = await orderRepo.findById(order.id);
+    const persistedLink = await customerOrderLinkRepo.find();
+    expect(order).to.deepEqual(persistedOrder);
+    expect(persistedLink).have.length(1);
+    const expected = {
+      id: 99,
+      customerId: existingCustomerId,
+      orderId: order.id,
+    };
+    expect(toJSON(persistedLink[0])).to.deepEqual(toJSON(expected));
+  });
+
+  it('can find an instance via through model', async () => {
+    const order = await hasManyThroughRepo.create(
+      {
+        description: 'an order hasManyThrough',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    const notMyOrder = await orderRepo.create({
+      description: "someone else's order desc",
+    });
+
+    const orders = await hasManyThroughRepo.find();
+
+    expect(orders).to.not.containEql(notMyOrder);
+    expect(orders).to.deepEqual([order]);
+  });
+
+  it('can find instances via through models', async () => {
+    const order1 = await hasManyThroughRepo.create(
+      {
+        description: 'group 1',
+      },
+      {
+        throughData: {id: 99},
+      },
+    );
+    const order2 = await hasManyThroughRepo.create(
+      {
+        description: 'group 2',
+      },
+      {
+        throughData: {id: 98},
+      },
+    );
+    const orders = await hasManyThroughRepo.find();
+
+    expect(orders).have.length(2);
+    expect(orders).to.deepEqual([order1, order2]);
+    const group1 = await hasManyThroughRepo.find({
+      where: {description: 'group 1'},
+    });
+    expect(group1).to.deepEqual([order1]);
+  });
+
+  it('can delete an instance then delet the through model', async () => {
+    await hasManyThroughRepo.create(
+      {
+        description: 'customer 1',
+      },
+      {
+        throughData: {id: 98},
+      },
+    );
+    const anotherHasManyThroughRepo = hasManyThroughFactory(
```

Comment:
hmm, how to make sure the customer with id as `existingCustomerId + 1` exists?

Reply:
I don't think we check the existence of the source key in any relation repositories, or it would need to access the database again.